src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs

```diff
@@ -0,0 +1,85 @@
+﻿using Files.App.Helpers;
+using Files.Shared.Enums;
+using System;
+
+namespace Files.App.Filesystem
+{
+    public class FileSystemProgress
+    {
+        private readonly IProgress<FileSystemProgress>? progress;
+        private readonly IntervalSampler sampler;
+        private FileSystemStatusCode? status;
+        private bool criticalReport;
+        private bool enumerationCompleted;
+
+        public FileSystemStatusCode? Status
+        {
+            get => status;
+            set
+            {
+                if (status != value)
+                {
+                    criticalReport = true;
+                }
+                status = value;
+            }
+        }
+        public string? FileName { get; set; }
+        public long TotalSize { get; set; }
+        public long ProcessedSize { get; set; }
+        public long ItemsCount { get; set; }
+        public long ProcessedItemsCount { get; set; }
+        public DateTimeOffset StartTime { get; }
+        public DateTimeOffset CompletedTime { get; private set; }
+        public bool EnumerationCompleted
+        {
+            get => enumerationCompleted;
+            set
+            {
+                if (enumerationCompleted != value)
+                {
+                    criticalReport = true;
+                }
+                enumerationCompleted = value;
+            }
+        }
+        /// <summary>
+        /// Only used when detailed count isn't available.
+        /// </summary>
+        public float? Percentage { get; set; }
+
+        public FileSystemProgress(IProgress<FileSystemProgress>? progress, int samplerInterval = 100)
+        {
+            this.StartTime = DateTimeOffset.Now;
+            this.progress = progress;
+            this.sampler = new(samplerInterval);
+        }
+
+        public void Report(float? percentage = null)
+        {
+            Percentage = percentage;
+            if ((EnumerationCompleted && (ProcessedItemsCount == ItemsCount || ProcessedSize == TotalSize)) ||
+                (percentage is float f && MathF.Abs(f - 100f) <= float.Epsilon))
+            {
+                if (status is FileSystemStatusCode.InProgress or null)
```

Comment:
This conditions are a mouthful to understand and the two `if`s can be merged together.
Worth putting it inside a method.